### PR TITLE
Add unit test for the the adjoint of the whole CAAR functor

### PR DESCRIPTION
**Pull request by bartgol**
Test the adjoint of the whole CAAR functor, inlcuding halo exchange and post-exchange kernel.

---

The current `run_JtV` function only computes the adjoint of local calculations. In a typical run, we also need the adjoint of the halo exchange as well as the post-exchange kernel (which simply sets some boundary conditions).

The tests are currently not passing. I'm creating the PR so I can ask for help from copilot, pointing it to the output of the failing CI tests.

This PR is built on top of #40, so it should be merged after that.

### CODE REVIEW

**Comment by bartgol** [comment]
@copilot The CI failed the unit test for the CAAR functor adjoint. Can you take a look? Some context:

- the test verifies that <x^1,lambda^1> = <x^0,lambda^0>, where X is the fwd state, while lambda is the adjoint state. We random init x^0 and lambda^1, and advance the two (x fwd, lambda bwd).
- the test case `caar_adjoint` is testing a sequence of operations that should be the adjoint of `caar.run(..)`. The current test assumes that the adjoint is made of
  1. the `caar.run_JtV_surf_bc` functor
  2. a halo exchange of the adjoint var
  3. the `caar.run_JtV` functor 
-  I verified that running just step 1 or just step 3 makes the test pass.
- I tried to disable the limiter in CAAR (which has no adjoint counterpart yet), but the test still fails.

**Comment by bartgol** [comment]
@copilot I removed the limiter call, but it did not help. That was not the issue (as I wrote in the previous message). Recap:

- if I only run the fwd/adj version of the pre-exchange loop, it works
- if I only run the fwd/adj version of the post-exchange loop, it works

The problem seems to arise when I run the fwd/adj version of the bex code. Can you investigate?

**Comment by bartgol** [comment]
Rebased on master after mergin #40 . Once CI passes again, I'll integrate.

File: components/homme/src/share/cxx/CaarFunctor_def.hpp [hunk 1]
```diff
@@ -72,61 +68,50 @@ setup(const ElementsST<ST> &elements, const TracersST<ST> &tracers,
...
 }
 
 template<typename ST>
 int CaarFunctorST<ST>::requested_buffer_size () const {
-  assert (m_caar_impl);
-  return m_caar_impl->requested_buffer_size();
+  assert (is_setup);
+  auto impl = std::any_cast<CaarFunctorImplST<ST>>(&m_caar_impl);
+  return impl->requested_buffer_size();
 }
```
[the remark this answers is not in the document]
Comment (bartgol) [reply components/homme/src/share/cxx/CaarFunctor_def.hpp:84]: num_elems is not known until setup has been called and the elements structures have been set.
